units/distro_overallocated_terminations.go

```diff
@@ -0,0 +1,131 @@
+)
+
+const (
+	hostDrawdownJobName = "host-drawdown"
+
+	// if we need to drawdown hosts, we want to catch as many hosts as we can that are between jobs
+	idleTimeDrawdownCutoff = 5 * time.Second
+)
+
+func init() {
+	registry.AddJobType(hostDrawdownJobName, func() amboy.Job {
+		return makeHostDrawdownJob()
+	})
+}
+
+type DrawdownInfo struct {
+	DistroID     string
+	NewCapTarget int
+}
+
+type hostDrawdownJob struct {
+	idleHostJob
+
+	DrawdownInfo DrawdownInfo
+}
+
+func makeHostDrawdownJob() *hostDrawdownJob {
+	j := &hostDrawdownJob{
+		idleHostJob{
+			Base: job.Base{
+				JobType: amboy.JobType{
+					Name:    hostDrawdownJobName,
+					Version: 0,
+				},
+			},
+		},
+		DrawdownInfo{},
+	}
+	j.SetDependency(dependency.NewAlways())
+	j.SetPriority(2)
+
+	return j
+}
+
+func NewHostDrawdownJob(env evergreen.Environment, drawdownInfo DrawdownInfo, id string) amboy.Job {
+	j := makeHostDrawdownJob()
+	j.DrawdownInfo = drawdownInfo
+	j.env = env
+	j.SetID(fmt.Sprintf("%s.%s", hostDrawdownJobName, id))
+	return j
+}
+
+func (j *hostDrawdownJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	if j.env == nil {
+		j.env = evergreen.GetEnvironment()
+	}
+
+	if j.settings == nil {
+		j.settings = j.env.Settings()
+	}
+
+	// get currently existing hosts, in case some hosts have already been terminated elsewhere
+	existingHostCount, err := host.CountRunningHosts(j.DrawdownInfo.DistroID)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "database error getting existing hosts by Distro.Id"))
+		return
+	}
+	idleHosts, err := host.IdleHostsWithDistroID(j.DrawdownInfo.DistroID)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "database error getting idle hosts by Distro.Id"))
+		return
+	}
+	drawdownTarget := existingHostCount - j.DrawdownInfo.NewCapTarget
+
+	for _, idleHost := range idleHosts {
+		if drawdownTarget <= 0 {
+			break
+		}
+		j.AddError(j.checkAndTerminateHost(ctx, &idleHost, &drawdownTarget))
+	}
+
+	grip.Info(message.Fields{
+		"id":                   j.ID(),
+		"job_type":             hostDrawdownJobName,
+		"distro_id":            j.DrawdownInfo.DistroID,
+		"num_idle_hosts":       len(idleHosts),
+		"num_terminated_hosts": j.Terminated,
+		"terminated_hosts":     j.TerminatedHosts,
+	})
+}
+
+func (j *hostDrawdownJob) checkAndTerminateHost(ctx context.Context, h *host.Host, drawdownTarget *int) error {
+
+	_, idleTime, err := j.getTimesFromHost(ctx, h)
+
+	if err != nil {
+		return errors.Wrapf(err, "error getting communication and idle time from '%s'", h.Id)
+	}
+
+	idleThreshold := idleTimeDrawdownCutoff
+	if h.RunningTaskGroup != "" {
+		idleThreshold = idleTaskGroupHostCutoff
+	}
+
+	if idleTime > idleThreshold {
+		terminateReason := fmt.Sprintf("host is being terminated due to temporary overallocation of hosts")
+		(*drawdownTarget)--
+		j.Terminated++
+		j.TerminatedHosts = append(j.TerminatedHosts, h.Id)
+		return j.env.RemoteQueue().Put(ctx, NewHostTerminationJob(j.env, h, false, terminateReason))
```

Comment:
Could you explain the logic of this job a bit? We currently terminate hosts if they have been idle more than N minutes. Is doing this different from reducing the idle minutes?

Separately, I remember when we discussed this that we discussed an approach of decommissioning running hosts when hosts were overallocated, not terminating idle hosts, since idle hosts were already about to be terminated.

Could you add a comment to the PR explaining in general terms how this change works?

Reply:
> is doing this different from reducing the idle minutes
Yes. This only goes into effect for a particular distro when that distro is very overallocated, and it has safeguards to make sure that hosts will only be terminated until the distro is near a host queue ratio of 1. I go over some of the reasons for this in detail [here](https://hhoke.github.io/evergreen_task_analysis/2021-03-22.html#startup-considerations-for-dynamic-distros). In a nutshell, when the host queue ratio is very low, the probability that a given idle host will have to be terminated after N minutes is very high, so we can skip the waiting and save money. When the host queue ratio is very high, this probability is very low, so we should be more generous about idle termination. Switching behavior on the host queue ratio value also avoids the problem of worrying about getNextTask performance mismatch, because if we're overallocated we don't want hosts we're about to terminate to pick up tasks.